quinn-proto/src/crypto/rustls.rs

```diff
@@ -260,47 +311,141 @@ impl crypto::ClientConfig for rustls::ClientConfig {
             next_secrets: None,
             inner: rustls::quic::Connection::Client(
                 rustls::quic::ClientConnection::new(
-                    self,
+                    self.inner.clone(),
                     version,
-                    server_name
-                        .try_into()
-                        .map_err(|_| ConnectError::InvalidDnsName(server_name.into()))?,
+                    ServerName::try_from(server_name)
+                        .map_err(|_| ConnectError::InvalidServerName(server_name.into()))?
+                        .to_owned(),
                     to_vec(params),
                 )
                 .unwrap(),
             ),
+            suite: Suite {
+                suite: self.initial.suite,
+                quic: self.initial.quic,
+            },
         }))
     }
 }
 
-impl crypto::ServerConfig for rustls::ServerConfig {
+impl TryFrom<rustls::ClientConfig> for QuicClientConfig {
+    type Error = NoInitialCipherSuite;
+
+    fn try_from(inner: rustls::ClientConfig) -> Result<Self, Self::Error> {
+        Ok(Self {
+            initial: initial_suite_from_provider(inner.crypto_provider())
+                .ok_or(NoInitialCipherSuite(()))?,
+            inner: Arc::new(inner),
+        })
+    }
+}
+
+/// The configured crypto provider does not support the QUIC initial cipher suite
+#[derive(Debug)]
+pub struct NoInitialCipherSuite(());
+
+impl std::fmt::Display for NoInitialCipherSuite {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        f.write_str("no initial cipher suite found")
+    }
+}
+
+impl std::error::Error for NoInitialCipherSuite {}
+
+/// A QUIC-compatible TLS server configuration
+///
+/// Can be constructed via [`ServerConfig::with_single_cert()`][single] or by using the
+/// [`TryFrom`] implementation with a custom [`rustls::ServerConfig`].
```

Comment:
We should document requirements on custom `rustls::ServerConfig`s: that TLS 1.3 be enabled exclusively, and that the early data limit be either `0` if 0-RTT is unwanted or `u32::MAX` otherwise.

Reply:
Extended the docstring here.